File: 백준/Platinum/8462. 배열의 힘/배열의 힘.cc
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

int n, m, arr[100001], freq[1000001];
ll answers[100001];
ll answer;
pair<int, pair<int, int>> q[100001];

void add(int cur) {
    answer -= 1LL * freq[arr[cur]] * freq[arr[cur]] * arr[cur];
    freq[arr[cur]]++;
    answer += 1LL * freq[arr[cur]] * freq[arr[cur]] * arr[cur];
}

void remove(int cur) {
    answer -= 1LL * freq[arr[cur]] * freq[arr[cur]] * arr[cur];
    freq[arr[cur]]--;
    answer += 1LL * freq[arr[cur]] * freq[arr[cur]] * arr[cur];
}

void process() {
    cin >> n >> m;
    for (int i = 1; i <= n; ++i) {
        cin >> arr[i];
    }
    for (int i = 1; i <= m; ++i) {
        int l, r;
        cin >> l >> r;
        q[i] = {i, {l, r}};
    }
    int sqrts = sqrt(n);
    sort(q + 1, q + m + 1, [&](pair<int, pair<int, int>> a, pair<int, pair<int, int>> b) {
        if (a.second.first / sqrts != b.second.first / sqrts) {
            return a.second.first / sqrts < b.second.first / sqrts;
        }
        return a.second.second < b.second.second;
    });
    auto& cur = q[1];
    int l = cur.second.first;
    int r = cur.second.second;
    for (int i = l; i <= r; ++i) {
        add(i);
    }
    answers[cur.first] = answer;

    for (int i = 2; i <= m; ++i) {
        while (l > q[i].second.first) {
            add(--l);
        }
        while (r < q[i].second.second) {
            add(++r);
        }
        while (l < q[i].second.first) {
            remove(l++);
        }
        while (r > q[i].second.second) {
            remove(r--);
        }
        answers[q[i].first] = answer;
    }
    for (int i = 1; i <= m; ++i) {
        cout << answers[i] << "\n";
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int o = 1;
//    cin >> o;
    while (o--) {
        process();
    }
    return 0;
}

File: 백준/Platinum/8462. 배열의 힘/배열의 힘.java
import static java.lang.Integer.parseInt;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Main {

    static class Pair {

        int id;
        int source;
        int dest;
        int comp;

        public Pair(final int id, final int source, final int dest, final int comp) {
            this.id = id;
            this.source = source;
            this.dest = dest;
            this.comp = comp;
        }
    }

    static int[] arr = new int[100_001];
    static long[] answers = new long[100_001];
    static int[] cnt = new int[100_000_1];
    static long answer = 0;
    static List<Pair> queries = new ArrayList<>();

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] s1 = br.readLine().split(" ");
        final var n = parseInt(s1[0]);
        final var m = parseInt(s1[1]);

        int sqrt = (int) Math.sqrt(n);

        final var s2 = br.readLine().split(" ");
        for (int i = 1; i <= n; i++) {
            arr[i] = parseInt(s2[i - 1]);
        }

        for (int i = 0; i < m; i++) {
            final var s = br.readLine().split(" ");
            final var src = parseInt(s[0]);
            queries.add(new Pair(i, src, parseInt(s[1]), src / sqrt));
        }

        queries.sort(
                (o1, o2) -> {
                    if (o1.comp != o2.comp) {
                        return Integer.compare(o1.comp, o2.comp);
                    }
                    return Integer.compare(o1.dest, o2.dest);
                }
        );

        Pair cur = queries.get(0);
        int src = cur.source;
        int dest = cur.dest;

        for (int i = src; i <= dest; i++) {
            add(arr[i]);
        }
        answers[cur.id] = answer;

        for (int i = 1; i < m; i++) {
            cur = queries.get(i);
            while (src > cur.source) {
                src--;
                add(arr[src]);
            }

            while (dest < cur.dest) {
                dest++;
                add(arr[dest]);
            }

            while (src < cur.source) {
                remove(arr[src]);
                src++;
            }

            while (dest > cur.dest) {
                remove(arr[dest]);
                dest--;
            }
            answers[cur.id] = answer;
        }

        final var sb = new StringBuilder();
        for (int i = 0; i < m; i++) {
            sb.append(answers[i]).append("\n");
        }
        System.out.print(sb);
    }

    private static void add(int cur) {
        answer -= (long) (cnt[cur]) * (cnt[cur]) * cur;
        cnt[cur]++;
        answer += (long) cnt[cur] * cnt[cur] * cur;
    }

    private static void remove(int cur) {
        answer -= (long) (cnt[cur]) * (cnt[cur]) * cur;
        cnt[cur]--;
        answer += (long) cnt[cur] * cnt[cur] * cur;
    }
}
